gb project importer

avidal opened the pull request:
### What does this do / why do we need it?

This adds a project importer for [gb](getgb.io), one more step towards converting all existing dependency managers.

### What should your reviewer look out for in this PR?

This was based on a combination of the [PR that added the glide importer](#380) plus the code in `glide` that added the `gb` importer: https://github.com/Masterminds/glide/blob/gps-integration/gb/gb.go

`gb` (well, more particularly, the `gb-vendor` plugin) doesn't save the tag if you fetch by tag, instead it fills in the `branch` key for the dependency with "HEAD", same if you fetch by a specific revision. This PR tries to infer a constraint based on just the revision, but I'm not sure if that's the appropriate behavior or not.

references #186

avidal commented:
I signed it!

avidal commented:
Hm. I rebased to clean up my commits and now I'm blocked on the CLA check again.

I signed it!

avidal commented:
Switched to only blessed repositories for the unit tests, as well as added an integration test (thanks @carolynvs!)

avidal commented:
Hmm. Both this PR and my other one #829 are waiting on the cla/google check since Friday..

sdboyer commented:
would be great if we could get a review from @davecheney on this - marking him as a reviewer 🙏

carolynvs commented on cmd/dep/gb_importer.go:
```diff
@@ -0,0 +1,169 @@
+	var revision = gps.Revision(pkg.Revision)
+
+	// See if we can get a version from that constraint
+	version, err := lookupVersionForLockedProject(pc.Ident, revision, revision, i.sm)
```
The second parameter `c gps.Constraint` should be `nil`, instead of passing in the revision again.

The purpose of the constraint parameter is to help pick the appropriate tag when a revision has multiple tags, by filtering with the semver constraint from the manifest (which we don't have with gb). To be clear, it doesn't cause the function to fail, but we know that since it's a revision, it's never going to help in that case. 😀

carolynvs commented on cmd/dep/testdata/gb/golden.txt:
```diff
@@ -0,0 +1,6 @@
+Detected gb manifest file...
+Converting from gb manifest...
+  Trying master (a0196ba) as initial lock for imported dep github.com/sdboyer/deptestdos
```
As noted above, we would like to be consistent in how these are logged and have the constraints listed before the locked projects.

carolynvs commented on cmd/dep/gb_importer.go:
```diff
@@ -0,0 +1,169 @@
+
+// gbManifest represents the manifest file for GB projects
+type gbManifest struct {
+	Version      int            `json:"version"`
```
Is this used anywhere?

carolynvs commented on cmd/dep/gb_importer.go:
```diff
@@ -0,0 +1,169 @@
+	// Branch may be HEAD or an actual branch. In the case of HEAD, that means
+	// the user vendored a dependency by specifying a tag or a specific revision
+	// which results in a detached HEAD
+	Branch string `json:"branch"`
```
It doesn't look like Branch is being used. If it's set to a real branch name that we can find in the repo, should it be used as the constraint in the manifest?

avidal replied:
I added support for using the branch as the initial constraint (if it's not HEAD); but I don't have a solid repository to use in the tests. Ideally there'd be a blessed test repository with two branches, then I can have a test that pins to a revision that is not on the master branch.

carolynvs replied:
Would https://github.com/carolynvs/deptest work for your test? It has a master and v2 branch, and it's "blessed" for using in dep's tests.

avidal replied:
I'll add a test that uses it and see how the tooling responds. Thanks.

carolynvs commented on cmd/dep/gb_importer.go:
```diff
@@ -0,0 +1,169 @@
+	// Path is used when the actual used package exists in a subpath
+	// For importing purposes, this isn't necessary as dep will figure
+	// out which packages are actually being used during resolving
+	Path string `json:"path,omitempty"`
```
Since we are ignoring `Path`, and don't print a warning about the fact that we ignore it, we can just remove it. 😄

carolynvs commented on cmd/dep/testdata/harness_tests/init/gb/case1/final/Gopkg.toml:
```diff
@@ -0,0 +1,10 @@
+  version = "1.0.0"
+
+[[constraint]]
+  branch = "master"
```
Right now, we are seeing "branch = master" here because the revision is the tip of master (which `lookupVersionForLockedProject` identifies). I am thinking that if the revision wasn't the tip however, we'd still like the constraint to be master. If we end up adding logic for using the branch from gb, using an earlier revision from master would be one way to test that out.

sdboyer commented:
how are we doing with this?

avidal commented:
@sdboyer @carolynvs Thanks for the feedback; been really busy now that I'm back at the office but I'll take a stab at addressing the comments once I get a chance.

avidal commented:
@carolynvs @sdboyer I just pushed a handful of commits that should address all of the feedback. Good call on using a non-tip revision of deptestdos, the test is much better now.

avidal commented:
Rebased to a single commit with all PR feedback addressed.

carolynvs commented on cmd/dep/gb_importer.go:
```diff
@@ -0,0 +1,170 @@
+	// But if the branch field is not "HEAD", we can use that as the initial constraint
+	var constraint gps.Constraint
+	if pkg.Branch != "HEAD" {
+		constraint = gps.NewBranch(pkg.Branch)
```
This method is missing a conditional branch that uses the constraint if a branch is set, and otherwise uses the result of `sm.InferConstraint`. Right now, constraint is set to master, then immediately discarded a few lines down.

Here are the steps we need:
1. Default the constraint to the branch if possible.
1. Use that constraint to lookup the version to put in the lock.
1. If the constraint is empty, try to infer it from the locked version.

avidal replied:
Ah, I misunderstood some of the functions I was calling. I'll rejigger it.

carolynvs commented on cmd/dep/testdata/gb/golden.txt:
```diff
@@ -0,0 +1,6 @@
+Detected gb manifest file...
+Converting from gb manifest...
+  Using ^2.0.0 as initial constraint for imported dep github.com/sdboyer/deptestdos
```
Once the constraint logic is fixed, this will change to `Using master as initial constraint`.

carolynvs commented on cmd/dep/gb_importer.go:
```diff
@@ -0,0 +1,170 @@
+		if err != nil {
+			return nil, nil, err
+		}
+		manifest.Constraints[pc.Ident.ProjectRoot] = gps.ProjectProperties{Source: pc.Ident.Source, Constraint: pc.Constraint}
```
In the gb manifest file, is it possible that the import path is a sub-package? e.g. the manifest has an entry for for "github.com/pkg/foo/cats" and "github.com/pkg/foo/dogs".

If so, then your importer will need to check if the root project has already been added, and skip subsequent imports for the same root package. For a reference, see how the [godep importer handles this scenario](https://github.com/golang/dep/blob/c300e1cdb5ea2218a73f3db1929dd1f44d9b03bf/cmd/dep/godep_importer.go#L103).

avidal replied:
Yes, I believe that the gb manifest has a "importpath" which contains a subpackage, and then the combination of "repository" and "path" points to the source for that specific subpackage. I dropped the path in the importer because dep will handle that already; and I misunderstood `gps.ProjectIdentifier`. I'll check the godep importer.

carolynvs replied:
I dorked this up for the glide importer, just opened #898 to fix. 😊

avidal commented:
Just wanted to give a huge 👍 to @carolynvs for being so thorough in the reviews. dep is a bit of a beast, and there are plenty of opportunities to misconstrue the architecture in an importer so having such a thorough maintainer for the subsystem is a huge help. Thanks again!

sdboyer commented:
@avidal you are correct, @carolynvs is awesome 😄 

thank you for sticking with it! there's obviously a lot of change in dep right now, which has definitely undermined our ability to generate good documentation - and what documentation we have is largely oriented towards users, not contributors.

if you feel like you've gained some valuable insights through this process that might be helpful for other new contributors, please please do share, and we'll see if we can turn those into docs, too 😄

avidal commented:
Hmm, the appveyor build failed because of an unrelated change. Is that expected?

avidal commented on cmd/dep/gb_importer.go:
```diff
@@ -0,0 +1,205 @@
+		}
+	}
+
+	// If there's *still* no constraint, set the constraint to the revision
```
Is this the proper behavior? Should there just not be a constraint if there's no branch specified and no tag that points to the revision?

carolynvs commented on cmd/dep/gb_importer.go:
```diff
@@ -156,6 +157,11 @@ func (i *gbImporter) convertOne(pkg gbDependency) (pc gps.ProjectConstraint, lp
+		So, the order of operations to convert a single dependency in a gb manifest is:
+			- Find a specific version for the revision (and branch, if set)
+			- If there's a branch available, use that as the constraint
+			- If there's no branch, but we found a version from step 1, use the version as the constraint
```
The inline doc is very helpful! 👍

carolynvs commented on cmd/dep/testdata/harness_tests/init/gb/case1/initial/src/gbtest/main.go:
```diff
@@ -0,0 +1,19 @@
+package main
+
+import (
+	cdt "github.com/carolynvs/deptest"
```
You don't have to change what you have here. I just wanted to share another pattern for making tests for dep:

```golang
package main

import (
  _ "github.com/carolynvs/deptest"
  _ "github.com/sdboyer/deptest"
)

func main() {}
```

carolynvs commented on cmd/dep/gb_importer_test.go:
```diff
@@ -0,0 +1,201 @@
+	}
+}
+
+func TestGbConfig_Convert_BadInput(t *testing.T) {
```
Would you please split this into two tests? Or single test with subtests whichever you are more comfortable with. One for the empty package name and another for the empty revision.

carolynvs commented on cmd/dep/testdata/harness_tests/init/gb/case1/final/Gopkg.toml:
```diff
@@ -0,0 +1,20 @@
+[[constraint]]
+  branch = "master"
+  name = "github.com/carolynvs/deptest"
+  source = "https://github.com/carolynvs/deptest"
```
These `source` fields shouldn't be present, because they aren't forks or non-standard repo locations.

It looks like gb always has `pkg.Repository` set... We need to come up some way to tell when we should set `source`. 🤔 @sdboyer I assume that gps already has logic that does that but I'm not sure where or it it's exposed.

carolynvs commented:
@avidal Thank for for the head start! @michael-go discovered when he added support for gvt that it also works for gb-vendor. We were able to use some of your logic here in this PR and got support for both merged in #1149.
